controller/api/destination/external_endpoint_manager.go

```diff
@@ -0,0 +1,275 @@
+	}
+
+	// Store configuration for leader elector client. The leader elector will
+	// accept three callbacks. When a lease is claimed, the elector will mark
+	// the manager as a 'leader'. When a lease is released, the elector will set
+	// the isLeader value back to false.
+	em.lec = leaderelection.LeaderElectionConfig{
+		// When runtime context is cancelled, lock will be released. Implies any
+		// code guarded by the lease _must_ finish before cancelling.
+		ReleaseOnCancel: true,
+		Lock: &resourcelock.LeaseLock{
+			LeaseMeta: metav1.ObjectMeta{
+				Name:      leaseName,
+				Namespace: controllerNs,
+			},
+			Client: k8sAPI.Client.CoordinationV1(),
+			LockConfig: resourcelock.ResourceLockConfig{
+				Identity: hostname,
+			},
+		},
+		LeaseDuration: leaseDuration,
+		RenewDeadline: leaseRenewDeadline,
+		RetryPeriod:   leaseRetryPeriod,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(context.Context) {
+				em.Lock()
+				defer em.Unlock()
+				em.isLeader = true
+			},
+			OnStoppedLeading: func() {
+				em.Lock()
+				defer em.Unlock()
+				em.isLeader = false
+				em.log.Infof("%s released lease", hostname)
+			},
+			OnNewLeader: func(identity string) {
+				if identity == hostname {
+					em.log.Infof("%s acquired lease", hostname)
+				}
+			},
+		},
+	}
+
+	return em, nil
+}
+
+// Start will run the endpoint manager's processing loop and leader elector.
+//
+// The function will spawn two background tasks; one to run the leader elector
+// client that and one that will process updates applied by the informer
+// callbacks.
+func (em *EndpointManager) Start() {
+	// Create a parent context that will be used by leader elector to gracefully
+	// shutdown.
+	//
+	// When cancelled (either through cancel function or by having its done
+	// channel closed), the leader elector will release the lease and stop its
+	// execution.
+	ctx, cancel := context.WithCancel(context.Background())
+
+	go func() {
+		for {
+			// Block until a lease is acquired or a lease has been released
+			leaderelection.RunOrDie(ctx, em.lec)
+			// If the context has been cancelled, exit the function, otherwise
+			// continue spinning.
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+		}
+	}()
+
+	// Start a background task to process updates. When a shutdown signal is
+	// received over the manager's stop channel, it is propagated to the elector
+	// through the context object to ensure the elector task does not leak.
+	go func() {
+		for {
+			select {
+			case update := <-em.updates:
+				em.processUpdate(update)
+			case <-em.stop:
+				em.log.Info("Received shutdown signal")
+				// Propagate shutdown to elector through the context.
+				cancel()
+				return
+			}
+		}
+	}()
+}
+
+// enqueueUpdate will enqueue a service key to the updates buffer to be
+// processed by the endpoint manager. When a write to the buffer blocks, the
+// update is dropped.
+func (em *EndpointManager) enqueueUpdate(update string) {
```

Comment:
Are we going to introduce some more structured representation to this update or we are going to keep it as a string?

Reply:
Probably not. We could pass in a pointer to a service though instead of a string. I think this will become a bit clearer once we move on to reconciliation.